### PR TITLE
feat: add ability to add checks with app ids

### LINKED ISSUE #1467: GHBranchProtection using deprecated field in required_status_checks
**Describe the bug**

[GHBranchProtection](https://github.com/hub4j/github-api/blob/4e103a6b192ca88c7948457e2b21ea3e6be9de89/src/main/java/org/kohsuke/github/GHBranchProtection.java#L245) is using the `contexts` field in RequiredStatusChecks. [This field](https://docs.github.com/en/enterprise-server@2.22/rest/reference/repos#update-branch-protection--parameters) was [deprecated in GitHub 3.0](https://docs.github.com/en/enterprise-server@3.0/rest/reference/branches#update-branch-protection--parameters).

> contexts(array of strings) | Required. Deprecated: The list of status checks to require in order to merge into this branch. If any of these checks have recently been set by a particular GitHub App, they will be required to come from that app in future for the branch to merge. Use checks instead of contexts for more fine-grained control.

As a result, repositories using App-authorized checks may experience surprising behaviour if two or more GitHub Apps are posting to the same check and a branch protection rule is (re)created.

**To Reproduce**

Create a branch protection rule with the API which requires one or more checks. Submit the required check from multiple GitHub Apps. Merging is blocked for one of the Apps.

image

**Expected behavior**

Checks can come from any PAT or GitHub App by default.

The new `checks` array should be passed to allow configuration of the required GitHub App for a check.

### PR DESCRIPTION
**Pull request by tginiotis-at-work**
# Description

Fixes #1467

Add ability to receive and specify App IDs for required checks in branch protection rules.

# Before submitting a PR:

- [x] Changes must not break binary backwards compatibility. If you are unclear on how to make the change you think is needed while maintaining backward compatibility, [CONTRIBUTING.md](CONTRIBUTING.md) for details.
- [x] Add JavaDocs and other comments explaining the behavior. 
- [x] When adding or updating methods that fetch entities, add `@link` JavaDoc entries to the relevant documentation on https://docs.github.com/en/rest . 
- [x] Add tests that cover any added or changed code. This generally requires capturing snapshot test data. See [CONTRIBUTING.md](CONTRIBUTING.md) for details.
- [x] Run `mvn -D enable-ci clean install site` locally. If this command doesn't succeed, your change will not pass CI.
- [x] Push your changes to a branch other than `main`. You will create your PR from that branch.

# When creating a PR: 

- [x] Fill in the "Description" above with clear summary of the changes. This includes:
  - [x] If this PR fixes one or more issues, include "Fixes #<issue number>" lines for each issue. 
  - [x] Provide links to relevant documentation on https://docs.github.com/en/rest where possible. If not including links, explain why not.
- [x] All lines of new code should be covered by tests as reported by code coverage. Any lines that are not covered must have PR comments explaining why they cannot be covered. For example, "Reaching this particular exception is hard and is not a particular common scenario."
- [x] Enable "Allow edits from maintainers".

### CODE REVIEW

File: src/main/java/org/kohsuke/github/GHBranchProtection.java [hunk 1]
```diff
@@ -221,6 +221,54 @@ public boolean isEnabled() {
...
 
+    /**
+     * The type Check.
+     */
+    public static class Check {
+        @JsonProperty
+        private String context;
+
+        @JsonProperty
+        private Integer app_id;
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHBranchProtection.java:232]:
```suggestion
        private String context;
        private Integer appId;
```

File: src/main/java/org/kohsuke/github/GHBranchProtection.java [hunk 2]
```diff
@@ -221,6 +221,54 @@ public boolean isEnabled() {
...
+         * @param context
+         *            the context string of the check
+         * @param appId
+         *            the application ID the check is supposed to come from
+         */
+        public Check(String context, Integer appId) {
+            this.context = context;
+            this.app_id = appId;
+        }
+
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHBranchProtection.java:252]:
```suggestion
```

File: src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java [hunk 3]
```diff
@@ -62,11 +81,24 @@ public GHBranchProtectionBuilder addRequiredChecks(Collection<String> checks) {
...
     }
 
+    /**
+     * Add required checks gh branch protection builder.
+     *
+     * @param checks
+     *            the checks
+     * @return the gh branch protection builder
+     */
+    public GHBranchProtectionBuilder addRequiredChecksWithAppIds(GHBranchProtection.Check... checks) {
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java:97]:
```suggestion
    public GHBranchProtectionBuilder addRequiredChecks(GHBranchProtection.Check... checks) {
```

File: src/main/java/org/kohsuke/github/GHBranchProtection.java (hunk 2 above)
Comment (tginiotis-at-work) [reply src/main/java/org/kohsuke/github/GHBranchProtection.java:252]: If this has no constructor, how can the user use the `GHBranchProtectionBuilder.addRequiredChecksWithAppIds` method, which accepts this object?

Comment (bitwiseman) [reply src/main/java/org/kohsuke/github/GHBranchProtection.java]: Jackson is magic.  It will do the binding for you.

Comment (tginiotis-at-work) [reply src/main/java/org/kohsuke/github/GHBranchProtection.java]: I don't think Jackson is related to the use case I am referring to. Here is an example of the use-case:

https://github.com/hub4j/github-api/blob/3882581deaedb92366c5a7b62b69a26c41f1113e/src/test/java/org/kohsuke/github/GHBranchProtectionTest.java#L203-L204

Or do you mean the user should use json strings and then Jackson to instantiate objects to pass in as arguments to that method?

Comment (bitwiseman) [reply src/main/java/org/kohsuke/github/GHBranchProtection.java]: Ah, I see.  My mistake.

File: src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java [hunk 4]
```diff
@@ -50,8 +50,27 @@ public class GHBranchProtectionBuilder {
...
+     *            the checks
+     * @return the gh branch protection builder
+     */
+    @Deprecated
     public GHBranchProtectionBuilder addRequiredChecks(Collection<String> checks) {
-        getStatusChecks().contexts.addAll(checks);
+        if (!(getStatusChecks() instanceof StatusChecksDeprecated)) {
+            statusChecks = new StatusChecksDeprecated();
+        }
+        ((StatusChecksDeprecated) getStatusChecks()).contexts.addAll(checks);
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java:73]: If someone calls both `addRequiredChecks` and `addRequiredStatusChecks`, only the last call will win.  
How about something like this:
```suggestion
        getStatusChecks(StatusChecksDeprecated.class).contexts.addAll(checks);
```

File: src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java [hunk 5]
```diff
@@ -532,7 +564,7 @@ private Restrictions getRestrictions() {
 
     private StatusChecks getStatusChecks() {
         if (statusChecks == null) {
-            statusChecks = new StatusChecks();
+            statusChecks = new StatusChecksWithAppId();
         }
         return statusChecks;
     }
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java:570]: Suggest something like this:
```
    <T extends StatusCheck> T getStatusChecks(Class<T> clazz) {
        if (statusChecks == null) {
            try {
                statusChecks = clazz.getDeclaredConstructor().newInstance();
            } catch (Exception ignored) {
            }
        }
        if (!clazz.isInstance(statusChecks) {
            throw new GHException("Cannot use checks and context status checks");
        } 
        return (T) statusChecks;
    }
```

File: src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java [hunk 6]
```diff
@@ -50,8 +50,27 @@ public class GHBranchProtectionBuilder {
      *            the checks
      * @return the gh branch protection builder
      */
+    public GHBranchProtectionBuilder addRequiredChecksWithAppIds(Collection<GHBranchProtection.Check> checks) {
+        if (!(getStatusChecks() instanceof StatusChecksWithAppId)) {
+            statusChecks = new StatusChecksWithAppId();
+        }
+        ((StatusChecksWithAppId) getStatusChecks()).checks.addAll(checks);
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java:57]:
```suggestion
         getStatusChecks(StatusChecksWithAppId.class).checks.addAll(checks);
```

**Review by bitwiseman — CHANGES_REQUESTED** [review]
Minor tweaks.  Thanks!

File: src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java [hunk 7]
```diff
@@ -50,8 +50,27 @@ public class GHBranchProtectionBuilder {
      *            the checks
      * @return the gh branch protection builder
      */
+    public GHBranchProtectionBuilder addRequiredChecksWithAppIds(Collection<GHBranchProtection.Check> checks) {
```
[the remark this answers is not in the document]
Comment (tginiotis-at-work) [reply src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java]: changed

File: src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tginiotis-at-work) [reply src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java]: changed

File: src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java (hunk 4 above)
Comment (tginiotis-at-work) [reply src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java]: https://github.com/hub4j/github-api/pull/1844#discussion_r1633525350

File: src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java (hunk 5 above)
Comment (tginiotis-at-work) [reply src/main/java/org/kohsuke/github/GHBranchProtectionBuilder.java]: Apparently the deprecated API can be implemented in the new API. From [the docs](https://docs.github.com/en/enterprise-server@3.0/rest/reference/branches#update-branch-protection--parameters):
image

I underlined the same logic between the old and new API.

I.e. you need to omit the `app_id` field to do it the old way. I've added the `@JsonInclude(JsonInclude.Include.NON_NULL)` to instruct jackson to omit it in that case, and a comment explaining the special "app_id" values.

So I committed a change which allows the use of both the old and new user methods at the same time, preserving both their effects.

How does that look to you?

**Review by bitwiseman — APPROVED** [review]

**Pull request merged**